homework_12/homework_12.py

```diff
@@ -0,0 +1,167 @@
+            pass
+
+    phone_book = r_dict()
+
+    # stats: кількість записів
+    # add: додати запис
+    # delete <name>: видалити запис за іменем (ключем)
+    # list: список всіх імен в книзі
+    # show <name>: детальна інформація по імені
+
+    try:
+        if user_input == "add":
+            add_input = input("enter a name and number: ")
+            add_split = add_input.split()
+            name = add_split[0]
+            num = int(add_split[1])
+            add_sub = phone_book.get(name)
+            if add_sub == None:
+                phone_book[name] = num
+                print("The subscribers " + add_split[0] + " is added to the phone book")
+                ubd_dict(phone_book)
+
+            else:
+                print("a subscriber with the name " + name + " already exists, try another name")
+
+        elif user_input == "stats":
+            len_dct = str(len(phone_book))
+            print("there are " + len_dct + " subscribers in this phone book")
+
+        elif user_input == "list":
+            sub_list = list(phone_book.keys())
+            all_sub = ""
+            for key_dct in sub_list:
+                if all_sub == "":
+                    all_sub = key_dct
+                else:
+                    all_sub = all_sub + ', ' + key_dct
+            print("All subscribers : " + all_sub)
+
+        if phone_book == {}:
+            print("Sorry, phone book is empty or you entered the wrong request")
+
+        else:
+            if user_input == "show":
+                show_input = input("enter a name: ")
+                show_sub = phone_book.get(show_input)
+                if show_sub != None:
+                    str_num = str(phone_book.get(show_input))
+                    print("Name " + show_input + " number " + str_num)
+                else:
+                    print("The subscribers " + show_input + " does not exist")
+
+            elif user_input == "delete":
+                del_input = input("enter a name: ")
+                del_sub = phone_book.get(del_input)
+                if del_sub != None:
+                    print("The subscribers " + del_input + " removed from the phone book")
+                    del phone_book[del_input]
+                    ubd_dict(phone_book)
+                else:
+                    print("The subscribers " + del_input + " does not exist")
+
+    except ValueError:
+        print("Sorry invalid request, try again")
+
+
+
+# 2. Написати декоратор, який буде записувати в файл назву функції, яку він декорує, і писати час її виклику.
+
+
+
+def deco_time_and_name(func_to_deco):
+    def print_time_and_name(*args, **kwargs):
+        str_name = func_to_deco.__name__
+        name_f = "function name " + str_name
+        print(name_f)
+        str_time = time.ctime()
+        name_t = "start function time " + str_time
+        print(name_t)
+        rez = func_to_deco(*args, **kwargs)
+
+        try:
+            with open('data_func.json', 'x+') as file:
```

Comment:
The modifier `a+` will create a file if it doesn't exist and will add logs to already existing content instead of rewriting it all the time